### ISSUE TITLE
torch.std cpu kernel segfaults when sub iterator is used in TensorIterator

### ISSUE THREAD

**Issue by jjsjann123** [pytorch/pytorch#15302]
## To Reproduce

import torch
torch.randn(200000, 2).std(dim=0)

## Expected behavior

Segmentation fault (core dumped)

## Environment

root@89e22c180051:/workspace# python collect_env.py
Collecting environment information...
PyTorch version: 1.0.0a0+a68751c
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.3 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.9) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.0.176
GPU models and configuration:
GPU 0: Graphics Device
GPU 1: TITAN V

Nvidia driver version: 410.27
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_static_v7.a

Versions of relevant libraries:
[pip] numpy (1.15.3)
[pip] torch (1.0.0a0+6100123)
[pip] torchvision (0.2.1)
[conda] blas                      1.0                         mkl
[conda] magma-cuda90              2.3.0                         1    pytorch
[conda] mkl                       2019.0                      118
[conda] mkl-include               2019.0                      118
[conda] mkl_fft                   1.0.6            py36h7dd41cf_0
[conda] mkl_random                1.0.1            py36h4414c95_1
[conda] torch                     1.0.0a0+f19e831           <pip>
[conda] torch                     1.0.0a0+75908b5           <pip>
[conda] torch                     1.0.0a0+1c8d41a           <pip>
[conda] torch                     1.0.0a0+fb7e40b           <pip>
[conda] torch                     1.0.0a0+e4f02f5           <pip>
[conda] torch                     1.0.0a0+ea0b149           <pip>
[conda] torch                     1.0.0a0+5d36ed3           <pip>
[conda] torch                     1.0.0a0+e17f833           <pip>
[conda] torch                     1.0.0a0+794aea6           <pip>
[conda] torch                     1.0.0a0+872b8a1           <pip>
[conda] torch                     1.0.0a0+506ae51           <pip>
[conda] torch                     1.0.0a0+6100123           <pip>
[conda] torch                     1.0.0a0+a68751c           <pip>
[conda] torch                     1.0.0a0+992fc2b           <pip>
[conda] torch                     1.0.0a0+01eb62a           <pip>
[conda] torch                     1.0.0a0+e829a52           <pip>
[conda] torchvision               0.2.1                     <pip>

## Additional context

The new kernel was introduced in: https://github.com/pytorch/pytorch/commit/25110d61fb868dcae5f1aee6edae38c8bea74523

Segfault happens here: https://github.com/pytorch/pytorch/blob/dcd1685282510b38afca5e522237bf7d8f69561a/aten/src/ATen/native/cpu/Reduce.h#L82-L112

**Comment by jjsjann123** [comment]
Adding @soumith for visibility.

**Comment by colesbury** [comment]
cc @umanwizard

**Comment by colesbury** [comment]
@jjsjann123 pointed in slack that the line:

```c++
std::vector<optional<acc_t>> buffer{(unsigned)max_threads, optional<acc_t> {}}; 
```

calls the wrong overload due to the `{}` instead of `()`

**Comment by jjsjann123** [comment]
Thanks for pointing out the bracket/parenthesis issue.
It has been fixed in my pending PR, which is not yet merged to master: https://github.com/pytorch/pytorch/pull/14878/files#diff-09c064e553b0c84f03c8472836f62cfdR89

**Comment by mruberry** [comment]
This appears to be fixed on nightly.
